src/exec.cpp

```diff
@@ -1182,30 +1182,31 @@ static int exec_subshell_internal(const wcstring &cmd, wcstring_list_t *lst,
     if (lst != NULL && io_buffer.get() != NULL) {
         const char *begin = io_buffer->out_buffer_ptr();
         const char *end = begin + io_buffer->out_buffer_size();
-        if (split_output) {
-            const char *cursor = begin;
-            while (cursor < end) {
-                // Look for the next separator.
-                const char *stop = (const char *)memchr(cursor, '\n', end - cursor);
-                const bool hit_separator = (stop != NULL);
-                if (!hit_separator) {
-                    // If it's not found, just use the end.
-                    stop = end;
-                }
-                // Stop now points at the first character we do not want to copy.
-                const wcstring wc = str2wcstring(cursor, stop - cursor);
-                lst->push_back(wc);
+        const char sep = split_output ? '\n' : '\0';
 
-                // If we hit a separator, skip over it; otherwise we're at the end.
-                cursor = stop + (hit_separator ? 1 : 0);
-            }
-        } else {
-            // we're not splitting output, but we still want to trim off a trailing newline.
+        if (!split_output) {
+            // we're not splitting output on newlines, but we still want to trim off
```

Comment:
This and the next if statement could just be `if (!split_output && end != begin && end[-1] == '\n') {` (OCLint added a rather minor warning).